### ISSUE TITLE
\dim_to_decimal_in_bp:n is not reentrant

### ISSUE THREAD

**Issue by jfbu** [latex3/latex3#954]
`\dim_to_decimal_in_bp:n` according to its documentation "leaves the result, expressed in big points (bp) in the input stream" (with no units).

But if the input is expressed in `bp` it is not always the case that the produced dimension is the same as original from TeX point of view.

And a value produced on output, if used as input, will perhaps not produce the same output (again from TeX point of view).

This is demonstrated by this:
```latex
\documentclass{article}

\ExplSyntaxOn
\newcommand\converttobp[1]{\dim_to_decimal_in_bp:n{#1}bp}
\ExplSyntaxOff
\makeatother

\begin{document}

577.35455786bp (\number\dimexpr 577.35455786bpsp)

\verb|\converttobp{577.35455786bp}| gives \converttobp{577.35455786bp}
(\number\dimexpr\converttobp{577.35455786bp}sp)

\ifdim577.35455786bp=\converttobp{577.35455786bp}\else They are not equal as
\TeX\ dimensions\fi

\verb|\converttobp{\converttobp{577.35455786bp}}| gives
\converttobp{\converttobp{577.35455786bp}}
(\number\dimexpr\converttobp{\converttobp{577.35455786bp}}sp)
which is yet a third \TeX\ dimension.

\end{document} 
```

which produces

```
577.35455786bp (37979398sp)
\converttobp{577.35455786bp} gives 577.35454bp (37979397sp)
They are not equal as TeX dimensions
\converttobp{\converttobp{577.35455786bp}} gives 577.35452bp (37979396sp)
which is yet a third TeX dimension.
```

The same applies to the more general `\dim_to_decimal_in_unit:nn` utility.

The above example was not obtained by any systematic search but simply by random choices and stopping as soon an example was found.

As indicated in the discussion at #953 it is actually possible to produce conversion functions using `\dimexpr` and `\numexpr`  which have the missing properties.

The point here is that in my opinion if user asks for a dimension in a certain unit, then it should be the case that if input back it is considered the same dimension from the TeX point of view... but admittedly I have looked only quickly the `l3skip` section in `interface3` and do not know if there is another concept of dimension there not fully TeX one.

Working only with 16 digits floating point numbers can not easily achieve this though (in some cases TeX's input may need to consider up to 17 places *after* decimal mark). [i am here saying this somewhat tentatively and did not check in any way how LaTeX3 handles these things].

**Comment by jfbu** [comment]
My musings about l3fp reveal an obsession but, had I opened source3.pdf, I would have known l3skip basically wraps around `\dimexpr`.

Ok, so the definition is currently

```latex
\cs_new:Npn \dim_to_decimal_in_bp:n #1
    { \dim_to_decimal:n { ( #1 ) * 800 / 803 } }
```

and I propose according to https://github.com/latex3/latex3/issues/953#issuecomment-869027810 this alternate definition

```latex
\cs_new:Npn \dim_to_decimal_in_bp:n #1
    { \dim_to_decimal:n 
      {
         \__int_eval:w ( 2 * \__dim_eval:w #1 \__dim_eval_end: + 1 ) * 3600 / 7227 sp 
      }
    }
```

All non font dependent dimension units pre-defined by TeX and larger than the TeX pt (i.e. all except `sp`) can be handled this way. For those units larger than 2pt, one can drop the +1 and merge the 2 in the ratio. The ratios must be picked in reverse from TeX source:

```
<Scan for all other units and adjust cur val and f accordingly; goto done in the case of scaled points 484 >
if scan keyword ("in") then set conversion (7227)(100)
else if scan keyword ("pc") then set conversion (12)(1)
else if scan keyword ("cm") then set conversion (7227)(254)
else if scan keyword ("mm") then set conversion (7227)(2540)
else if scan keyword ("bp") then set conversion (7227)(7200)
else if scan keyword ("dd") then set conversion (1238)(1157)
else if scan keyword ("cc") then set conversion (14856)(1157)
else if scan keyword ("nd") then set conversion (685)(642)
else if scan keyword ("nc") then set conversion (1370)(107)
```

For example
```latex
\cs_new:Npn \dim_to_decimal_in_dd:n #1
    { \dim_to_decimal:n 
      {
         \__int_eval:w ( 2 * \__dim_eval:w #1 \__dim_eval_end: + 1 ) * 1157 / 2476 sp 
      }
    }
```

Regarding `ex` and `em` this is more challenging and costly as
- it would need to hook into font selection
- anyhow there is a problem of principle if those units end up to be less than `1pt`.

Unfortunately, not for all units is the output of conversion of `\maxdimen` usable as input.

For conversion to `bp`, it works with `\maxdimen`: the output is usable as input. Bur for conversion to `dd`, the output is `15312.025845` and `\dimexpr15312.025845dd\relax` triggers a dimension too large error.

In fact it is impossible to obtain `\maxdimen` from a dimension expressed in `dd`, as can be shown. The maximal obtainable dimension from a `dd` specification is `1073741822sp`, but `1073741823sp` is impossible to reach, beyond `1073741822sp` one can only trigger "Dimension too large" error from a `dd` unit dimension.

To obtain `\maxdimen` from an input in `dd` it would be necessary that this equation in an integer N has a solution:
```
1073741823<= N * 1238/1157 < 1073741824
```
but this translates into
```
1003488925.049273... <= N < 1003488925.983844911147....
```
which has no solution.

For the `bp` unit, this phenomenon does not happen as `16322.78954bp` is parsed by TeX internally into the same as `\maxdimen`.

The macros are guaranteed to produce an output which gives same TeX dimension as input **only** if the input is expressed (or express-able) in the corresponding unit. Here `\maxdimen` is not express-able in `dd` unit. Turns out the procedure in this case would then give the next `sp` (in the bounds for N above, the average of lower and upper bound turns out to be greater than `1003488925.5` and its rounding is thus increasing it to produce `1003488926`) and the Dimension too large is triggered if we use as input the output of `\maxdimen`.

Example with `bp`:
```latex
\documentclass{article}

\ExplSyntaxOn
% original
%\cs_new:Npn \dim_to_decimal_in_bp:n #1
%    { \dim_to_decimal:n { ( #1 ) * 800 / 803 } }
% ma version
% \strip@pt\dimexpr\numexpr3600*(2*\dimexpr#1\relax+1)/7227spbp
% traduction en LaTeX3
\cs_new:Npn \jfbu_dim_to_decimal_in_bp:n #1
    { \dim_to_decimal:n 
      {
         \__int_eval:w ( 2 * \__dim_eval:w #1 \__dim_eval_end: + 1 ) * 3600 / 7227 sp 
      }
    }
\newcommand\converttobp[1]{\jfbu_dim_to_decimal_in_bp:n{#1}bp}
\ExplSyntaxOff
\makeatother

\begin{document}

577.35455786bp (\number\dimexpr 577.35455786bpsp)

\verb|\converttobp{577.35455786bp}| gives \converttobp{577.35455786bp}
(\number\dimexpr\converttobp{577.35455786bp}sp)

\ifdim577.35455786bp=\converttobp{577.35455786bp}They are indeed the same
\TeX\ dimensions\fi.

More generally \verb|\ifdim| will always consider equal a dimension expressed
in bp units and the output of \verb|converttobp| as applied to this dimension.

\converttobp{\maxdimen}

\converttobp{\converttobp{\maxdimen}}

\end{document}
```

produces

```
577.35455786bp (37979398sp)
\converttobp{577.35455786bp} gives 577.35455bp (37979398sp)
They are indeed the same TEX dimensions.
More generally \ifdim will always consider equal a dimension expressed in
bp units and the output of converttobp as applied to this dimension.
16322.78954bp
16322.78954bp
```

**Comment by jfbu** [comment]
I forgot to simplify the fraction (because Knuth does not either). So:
```latex
\cs_new:Npn \dim_to_decimal_in_bp:n #1
    { \dim_to_decimal:n 
      {
         \__int_eval:w ( 2 * \__dim_eval:w #1 \__dim_eval_end: + 1 ) * 400 / 803 sp 
      }
    }
```

In this case of `bp`  or the other units `<2pt` it can be proven that if the original dimension `#1` has no exact equivalent in `bp` unit, then both `#1+1sp` and `#1-1sp` have such an exact equivalent and the procedure produces one of them (depending on the `#1` it may be `+1sp` or `-1sp`).

I have not thought this through seriously but I see it would be possible to modify the process to always produce a decimal D such that `Dbp` is either seen by TeX as the same dimension as original `#1`, or, if this is not possible as `#1-1sp`. This does not seem to have interest in case of `bp` as `\maxdimen` is representable exactly, but in the case of `dd` (which is also `<2pt`) this would produce an output decimal Dout such that `Dout bp` will be seen by TeX as `\maxdimen-1sp` rather than as a `\maxdimen+1sp` triggering Dimension too large.

The complication and added overhead may not be worth the effort. Already my proposal above to "fix" `\dim_to_decimal_in_bp:n` may be seen un-needed, so I will wait if feedback comes to think more about this.

**Comment by jfbu** [comment]
(Un)fortunately I did not consider the case of a negative `#1`. One must replace `+1` by `-1` for them in the conversion formula... this complicates things... again I will simply wait to see if there is interest before thinking this through. It is possible the overhead needed to check the sign may then be put to profit and help implementing the "safe choice" described in previous comment (which is only for units `<2pt` other units need more thinking about what one wants to do). After all @dbitouze may be right that this would need to be experimented in a package first.

**Comment by jfbu** [comment]
Here is the list of all decimals `X = 12.dd` such that current `\dim_decimal_to_bp:n{Xbp}` produces a `Y` such that TeX considers `Xbp` and `Ybp` to be different dimensions, to show this is not a rare phenomenon:
```
12.01 (790038, 790037)
12.04 (792011, 792010)
12.06 (793327, 793326)
12.08 (794643, 794642)
12.10 (795959, 795958)
12.12 (797274, 797273)
12.14 (798590, 798589)
12.15 (799247, 799246)
12.17 (800563, 800562)
12.19 (801879, 801878)
12.21 (803195, 803194)
12.23 (804510, 804509)
12.25 (805826, 805825)
12.28 (807799, 807798)
12.30 (809115, 809114)
12.32 (810431, 810430)
12.34 (811746, 811745)
12.36 (813062, 813061)
12.38 (814378, 814377)
12.39 (815035, 815034)
12.41 (816351, 816350)
12.43 (817666, 817665)
12.45 (818982, 818981)
12.47 (820298, 820297)
12.49 (821614, 821613)
12.52 (823587, 823586)
12.54 (824902, 824901)
12.56 (826218, 826217)
12.58 (827534, 827533)
12.60 (828850, 828849)
12.63 (830823, 830822)
12.65 (832138, 832137)
12.67 (833454, 833453)
12.69 (834770, 834769)
12.71 (836086, 836085)
12.73 (837401, 837400)
12.74 (838059, 838058)
12.76 (839374, 839373)
12.78 (840690, 840689)
12.80 (842006, 842005)
12.82 (843322, 843321)
12.84 (844637, 844636)
12.87 (846610, 846609)
12.89 (847926, 847925)
12.91 (849242, 849241)
12.93 (850557, 850556)
12.95 (851873, 851872)
12.97 (853189, 853188)
12.98 (853846, 853845)
```

Displayed as ``X (Xbp in scaled points, \dim_to_decimal_in_bp:n{Xbp}bp in scaled points)``.

**Comment by jfbu** [comment]
Mathematically here is how to obtain from a positive TeX dimension `T sp` the largest dimension D smaller or equal to `T sp` and exprimable in the unit `bp`,  respectively `dd`.

For `bp`, one computes the integer `N` equals to `ceil((T+1)*800/803)-1`. Then one uses `\the\dimexpr Nsp` to obtain a decimal `X`. Then `X bp` is the looked-for dimension. Equivalently, from the `N` one computes `U = floor(N*803/800)` and `Usp` is the largest dimension exprimable in the `bp` unit and at most equal to `Tsp`.

For `dd` one does exactly the same replacing `800/803` by `1157/1238`, respectively `803/800` by `1238/1157`.

For the other units, same.

**Comment by jfbu** [comment]
Regarding the smallest dimension D at least equal to `T sp` and exactly exprimable in the unit `bp`,  it is obtained as `Usp` with `U = floor(N*803/800)` and `N = ceil(T*800/803)`.  (here `T` is non negative).

Same for other units.

**Comment by blefloch** [comment]
Hi @jfbu.  Your points about `\dim_to_decimal_in_bp:n` are very relevant.  For other units, I don't see how `\dim_to_decimal_in_unit:nn` could detect that its second argument is a unit such as `1mm` or `1dd`, so I don't think there is anything we could do, except perhaps improve the rounding.  What do you think?

**Comment by jfbu** [comment]
Hi @blefloch my initial reaction is that I agree with you about `\dim_to_decimal_in_unit:nn`, this is also why I mentioned `em` and `ex` as delicate (to be honest, I don't know or remember the precise details of how TeX handles them).

By the way I don't quite understand the interface of this function, as its second argument is a dim. But from the name I would have thought the interface was `\dim_to_decimal_in_unit{<dim>}{bp or dd or mm or cm or...}` and that it would act as a dispatch to individual converters each in the style of `\dim_to_decimal_in_bp:n`. The name `\dim_ratio_as_decimal:nn` would be more revealing of what it actually does...

**Comment by jfbu** [comment]
I have a setup a [jfbu/texdimens](https//github.com/jfbu/texdimens) project.

Most of the [documentation](https://github.com/jfbu/texdimens/blob/master/texdimens/README.md) is done but I yet have to implement the math as described above. (current implementation idea is: as it will not be possible to do `ceil` and `floor` via `\numexpr` scaling operations without potential overflow, the rounding path (as described here for `bp`) will be used, then a second step will check if we are above or below target dimension to adjust to what is asked for; near `\maxdimen` this will be a bit problematic as testing may trigger "Dimension too large" so I probably will add "safe" variants which will filter out dimensions too near `\maxdimen`.)

Feedback welcome.

**Comment by jfbu** [comment]
My experimental package ([README](https://github.com/jfbu/texdimens/blob/master/texdimens/README.md))  has now reached a relatively stable status. I have not tested it really but it can contribute to illustrate what is at stake in converting tex dimension from one unit to another.

**Comment by RuixiZhang42** [comment]
The discussions here have impact on real-world applications. I noticed one such anomaly when maintaining the `ctex` bundle. The `ctex` document classes offer a `10.5bp` font size for body text, known as 五号 in simplified Chinese (five hao). The natural leading is 1.2 times the font size, i.e., `12.6bp`.

Of course, an input of `12.6bp` is internally represented by the integer `828850`. However, asking either `\dim_to_decimal_in_bp:n \baselineskip` or simply `\dim_to_decimal_in_bp:n { 12.6bp }` yields `12.59999`. But `12.59999bp` is internally `828849`, not the original `828850` (!!!). So the error accumulates and carries over *on every page*.

Knuth made sure that a `pt` to `sp` conversion followed by `sp` to `pt` decimal output would have the two `pt` internally identical. Well, the user decimal input `pt` may differ from the decimal output `pt`, but at least the two versions map to the same internal `sp`. Any units other than `pt` are the Wild West. If care were not taken during conversion, anything could happen.

**Comment by RuixiZhang42** [comment]
@blefloch There seems to be no progress made on this issue. I&nbsp;took the opportunity to write a detailed analysis. I&nbsp;also provided a mathematical proof on why the proposed method works. The whole document is three pages long (the proof is fairly detailed). Please feel free to incorporate my text and/or code into `l3skip`; they are released under LPPL Version&nbsp;1.3c.

@jfbu I&nbsp;simplified your proposed method a bit by dropping the `\numexpr`. Hope you don&rsquo;t mind `:)`.

```latex
% The content of this entire document is released under LPPL Version 1.3c.

\documentclass{article}
\usepackage{amsmath}

\makeatletter
\newcommand*\<{}%>
\def\<#1>{\leavevmode\hbox{$\langle\m@th$#1$\rangle\m@th$}}
\newcommand*\?[1]{\leavevmode\hbox{\normalfont\ttfamily#1}}
\makeatother

\begin{document}

\hbox to\hsize{\hss Ruixi Zhang, November~1, 2021}

\section{Preliminary}

Given a dimension input of ``\<$n$>\?{.}\<$d_1d_2\ldots d_{16}$>\?{bp}''
($n$ and $d_i$ are integers with $n\ge0$ and $0\le d_i\le9$),
\TeX\ will turn this into its internal representation of ``\<$X$>\?{sp}'', where
\[
X=\left\lfloor\Bigl\lfloor2^{16}\times\bigl(\hbox{\<$n$>\?{.}\<$d_1d_2\ldots d_{16}$>}\bigr)+\tfrac{1}{2}\Bigr\rfloor\times\frac{7227}{7200}\right\rfloor.
\]
To facilitate the formula above: \TeX\ first ``rounds'' the decimal
\<$n$>\?{.}\<$d_1d_2\ldots d_{16}$> to the nearest $n+f/2^{16}=(2^{16}n+f)/2^{16}$,
where $0\le f\le2^{16}$ is an integer representing the fractional part
(the case $f=2^{16}$ can happen, e.g., when we input \?{0.999993}).
The numerator $2^{16}n+f$ is called a ``scaled integer'' in \TeX's terminology.
In the second step, \TeX\ applies a ``conversion factor'' to $2^{16}n+f$.
This is done in 1.5-precision arithmetic and the final result is ``truncated''.
The conversion factor associated with \?{bp} is $7227/7200$---%
let's denote it by~$\delta=\delta_{\?{bp}}$.

\section{What should \?{\char`\\dim\char`\_to\char`\_decimal\char`\_in\char`\_bp:n} do?}

What \verb|\dim_to_decimal_in_bp:n| is supposed to do is to take
the input of \<$X$>\?{sp} and produce an output of a decimal
\<$n'$>\?{.}\<$d_1'd_2'd_3'd_4'd_5'$> (or the $2^{16}n'+f'$ equivalent).
This function \emph{must} guarantee that its output---%
when used as input ``\<$n'$>\?{.}\<$d_1'd_2'd_3'd_4'd_5'$>\?{bp}''---%
can convert back to exactly \<$X$>\?{sp}.

In other words, when given $X=\lfloor(2^{16}n+f)\delta\rfloor$, we seek a
$2^{16}n'+f'$ such that $X=\lfloor(2^{16}n'+f')\delta\rfloor$ too.
Note that the two scaled integers $2^{16}n+f$ and $2^{16}n'+f'$ are not
necessarily equal
(and by extension the two decimals \<$n$>\?{.}\<$d_1d_2\ldots d_{16}$>
and \<$n'$>\?{.}\<$d_1'd_2'd_3'd_4'd_5'$> are not necessarily equal).
We simply demand (and we must demand) that
$\lfloor(2^{16}n+f)\delta\rfloor=\lfloor(2^{16}n'+f')\delta\rfloor$.

\section{Current implementation and problems}

The current implementation of\/ \verb|\dim_to_decimal_in_bp:n| simply does
\[
2^{16}n'+f'\leftarrow\bigl\lfloor X\times\delta^{-1}+\tfrac{1}{2}\bigr\rfloor,
\]
where $\delta^{-1}=7200/7227=800/803$ is the hard-coded inverse of the
conversion factor, and the rounding comes from $\varepsilon$-\TeX's
``scaling'' operation.

The question is:
\[
X=\lfloor(2^{16}n+f)\delta\rfloor\overset{???}{=}\lfloor(2^{16}n'+f')\delta\rfloor
=\bigl\lfloor\lfloor X\times\delta^{-1}+\tfrac{1}{2}\rfloor\delta\bigr\rfloor.
\]
As the (many) counterexamples have shown, the above does \emph{not} always hold.
One such counterexample is \?{12.6bp}, where 
$X=\lfloor825754\times\delta\rfloor=828850$, but
$\bigl\lfloor\lfloor X\times\delta^{-1}+\tfrac{1}{2}\rfloor\delta\bigr\rfloor
=\lfloor825753\times\delta\rfloor=828849$.
By the way, $2^{16}n+f=825754$ translates to exactly \?{12.6},
while $2^{16}n'+f'=825753$ translates to \?{12.59999}.

\section{Proposed method and proof that it works}

So the current implementation fails. What can we do? Well, the proposed method
can be further simplified to
\begin{verbatim}
\cs_new:Npn \dim_to_decimal_in_bp:n #1
  {
    \dim_to_decimal:n
      {
%% We are already inside a dimen expression,
%% so no need to use another int expression.
        ( 2 \__dim_eval:w #1 \__dim_eval_end: + 1sp ) * 400 / 803
%% This is ( 2X + 1 ) * 400 / 803, where 400/803 = (1/2)(1/delta)
      }
  }
\end{verbatim}
which essentially does the following:
\[
2^{16}n'+f'\leftarrow\bigl\lfloor(2X+1)\times(\tfrac{1}{2}\delta^{-1})+\tfrac{1}{2}\bigr\rfloor,
\]
where $\tfrac{1}{2}\delta^{-1}=3600/7227=400/803$, and it is guaranteed that
for all integers $X=\lfloor N\delta\rfloor$, we have
\[
X=\bigl\lfloor\lfloor(2X+1)\times(\tfrac{1}{2}\delta^{-1})+\tfrac{1}{2}\rfloor\delta\bigr\rfloor.
\]

\noindent
{\itshape Proof.}
Let $Y$ denote the right-hand side, i.e.,
\[
Y=\bigl\lfloor\lfloor(2X+1)\times(\tfrac{1}{2}\delta^{-1})+\tfrac{1}{2}\rfloor\delta\bigr\rfloor
=\bigl\lfloor\lfloor X\delta^{-1}+\tfrac{1}{2}\delta^{-1}+\tfrac{1}{2}\rfloor\delta\bigr\rfloor.
\]
We shall prove that $Y=X$.

First, we remove the outer floor and obtain
\[
Y\le\lfloor X\delta^{-1}+\tfrac{1}{2}\delta^{-1}+\tfrac{1}{2}\rfloor\delta<Y+1.
\]
Then, we divide by~$\delta$ and derive
\begin{equation}
Y\delta^{-1}\le\lfloor X\delta^{-1}+\tfrac{1}{2}\delta^{-1}+\tfrac{1}{2}\rfloor<Y\delta^{-1}+\delta^{-1}.
\label{eq:fund}
\end{equation}
Statement~(\ref{eq:fund}) has the following visual representations:
\begin{center}
\setlength{\unitlength}{20pt}
\providecommand*\clap[1]{\hbox to0pt{\hss#1\hss}}
\begin{picture}(5,2)
\put(0,0){\line(1,0){5}}
\put(1.5,0.125){\line(0,-1){0.25}}
\put(1.5,-0.4375){\clap{$\scriptstyle n$}}
\put(3.5,0.125){\line(0,-1){0.25}}
\put(3.5,-0.4375){\clap{$\scriptstyle n+1$}}
\put(0.125,0.25){\line(0,-1){0.375}}
\put(0.125,0.375){\clap{$\scriptstyle Y\!\rlap{$\scriptstyle /\delta$}$}}
\put(1.625,0.25){\line(0,-1){0.375}}
\put(1.625,0.375){\clap{$\scriptstyle Y\!\rlap{$\scriptstyle /\delta+1/\delta$}$}}
\put(3.375,1.5){\vector(0,-1){1.5}}
\put(3.375,1.75){\clap{$\scriptstyle X\delta^{-1}+\frac{1}{2}\delta^{-1}+\frac{1}{2}$}}
\end{picture}%
\kern20pt
or%
\kern20pt
\begin{picture}(5,2)
\put(0,0){\line(1,0){5}}
\put(1.5,0.125){\line(0,-1){0.25}}
\put(1.5,-0.4375){\clap{$\scriptstyle n$}}
\put(3.5,0.125){\line(0,-1){0.25}}
\put(3.5,-0.4375){\clap{$\scriptstyle n+1$}}
\put(1.5,0.25){\line(0,-1){0.375}}
\put(1.5,0.375){\clap{$\llap{$\scriptstyle Y\!/$}\scriptstyle\delta\,$}}
\put(3,0.25){\line(0,-1){0.375}}
\put(3,0.375){\clap{$\scriptstyle Y\!\rlap{$\scriptstyle /\delta+1/\delta$}$}}
\put(1.625,1.5){\vector(0,-1){1.5}}
\put(1.625,1.75){\clap{$\scriptstyle X\delta^{-1}+\frac{1}{2}\delta^{-1}+\frac{1}{2}$}}
\end{picture}
\end{center}

\medskip

Because $X=\lfloor N\delta\rfloor$ for some integer $N\ge0$,
the number $X\delta^{-1}+\frac{1}{2}\delta^{-1}+\frac{1}{2}$ \emph{cannot}
get arbitrarily close to $n+1$ in the left picture,
\emph{nor} can it get arbitrarily close to~$n$ in the right picture,
where $n=\lfloor X\delta^{-1}+\frac{1}{2}\delta^{-1}+\frac{1}{2}\rfloor$.
In fact,
\[
\max_{X=\lfloor N\delta\rfloor}\Bigl\{\bigl(X\delta^{-1}+\tfrac{1}{2}\delta^{-1}+\tfrac{1}{2}\bigr)-n\Bigr\}=\frac{1603}{1606},
\]
and this maximum is attained if and only if $N\equiv0\pmod{800}$.
For example, $\?{25bp}\implies N=1638400=2048\times800\implies X=1644544$,
and $X\delta^{-1}+\frac{1}{2}\delta^{-1}+\frac{1}{2}=1638400+\frac{1603}{1606}$.
Similarly,
\[
\min_{X=\lfloor N\delta\rfloor}\Bigl\{\bigl(X\delta^{-1}+\tfrac{1}{2}\delta^{-1}+\tfrac{1}{2}\bigr)-n\Bigr\}=\frac{5}{1606},
\]
and this minimum is attained if and only if $N\equiv533\pmod{800}$.
For example, $\?{0.118bp}\implies N=7733=9\times800+533\implies X=7761$,
and $X\delta^{-1}+\frac{1}{2}\delta^{-1}+\frac{1}{2}=7733+\frac{5}{1606}$.

Next, let us look at the number $Y\delta^{-1}+\delta^{-1}$ in~(\ref{eq:fund}).
Because the conversion factor $\delta>1$, its inverse $\delta^{-1}<1$,
so we must have
\[
\lfloor Y\delta^{-1}+\delta^{-1}\rfloor=n=\lfloor X\delta^{-1}+\tfrac{1}{2}\delta^{-1}+\tfrac{1}{2}\rfloor.
\]
In the left picture, we must respect the strict inequality
\[
n<Y\delta^{-1}+\delta^{-1}\implies n-Y\delta^{-1}-\delta^{-1}<0.
\]
In the right picture, we have naturally
\[
Y\delta^{-1}\le n\implies n-Y\delta^{-1}-\delta^{-1}\ge-\delta^{-1}=-\frac{800}{803}.
\]

To summarize, we have shown that
\begin{equation}
\frac{5}{1606}\le X\delta^{-1}+\tfrac{1}{2}\delta^{-1}+\tfrac{1}{2}-n\le\frac{1603}{1606},
\label{eq:for-X}
\end{equation}
and
\begin{equation}
-\frac{800}{803}\le n-Y\delta^{-1}-\delta^{-1}<0.
\label{eq:for-Y}
\end{equation}
Adding (\ref{eq:for-X}) and~(\ref{eq:for-Y}) yields
\[
-\frac{145}{146}\le X\delta^{-1}-Y\delta^{-1}-\tfrac{1}{2}\delta^{-1}+\tfrac{1}{2}<\frac{1603}{1606}.
\]
Multiplying~$\delta$ yields
\[
-\frac{319}{320}\le X-Y-\tfrac{1}{2}+\tfrac{1}{2}\delta<\frac{1603}{1600}.
\]
Finally, rearranging the terms, we get
\[
-\frac{319}{320}+\frac{1}{2}-\frac{1}{2}\delta\le X-Y<\frac{1603}{1600}+\frac{1}{2}-\frac{1}{2}\delta,
\]
or equivalently,
\[
-\frac{799}{800}\le X-Y<1.
\]
But both $X$ and~$Y$ are integers, and so is their difference.
Therefore, it must be that $X-Y=0$, i.e., $X=Y$ as required.\nobreak\hfill QED

\end{document}
```

**Comment by jfbu** [comment]
@RuixiZhang42 Kuddos on devoting time on such a detailed math account, great! 
<sub>warning next paragraph is written for gentle fun Censors of Code Of Conducts and dictators of good taste controlling all aspects of expressions on internet, please take this into account before applying the severity of the (self-edicted) affirmative laws.</sub>
Perhaps it will make its way into university curricula at some point :-)  (even before mastering LaTeX3 syntax becomes a requirement of all CS majors making my prospect of acquiring the degree more distant yet, anyhow I am doomed as I never studied AOCP in my too aetherally mathematical youth)

I contributed, motivated by this issue,  a [`texdimens`](https://github.com/jfbu/texdimens) TeX package (which can be used by all flavors of TeX). Therein, some (very condensed) mathematical explanations are given within the [commented out parts of the source code](https://github.com/jfbu/texdimens/tree/master/texdimens/texdimens.tex).

<strike>As per the `\numexpr`, it was probably to avoid overflow, I tested 
```
\ExplSyntaxOn
\dim_to_decimal_in_bp:n \maxdimen
\ExplSyntaxOff
```
after having copied pasted your version without the `\numexpr` and I got this:
```
./ruixi.tex:25: Dimension too large.
<argument> ...m_eval:w \maxdimen \__dim_eval_end: 
                                                  +1sp)*400/803
l.25 \dim_to_decimal_in_bp:n \maxdimen
                                      
I can't work with sizes bigger than about 19 feet.
Continue and I'll use the largest value I can.
```
</strike>
<sub>I see you edited your comment while I was finishing my message...</sub>

Problems with `\maxdimen` are even more acute for some considerations with other units, as mentioned in [`texdimens` code README](https://github.com/jfbu/texdimens/blob/master/texdimens/README.md).

**Comment by RuixiZhang42** [comment]
@jfbu Yes, when using a particular `<unit>` (any TeX unit other than `sp` and `pt`), there must exist *inexpressible* dimensions `<X>sp`, in the sense that `X` can never come from an input of `<n>.<dd...d><unit>`. In fact, for each unit, we know exactly how many dimensions are inexpressible:
```latex
% The content of this entire document is released under LPPL Version 1.3c.

\documentclass{article}
\usepackage{amsmath}

\makeatletter
\newcommand*\<{}%>
\def\<#1>{\leavevmode\hbox{$\langle\m@th$#1$\rangle\m@th$}}
\newcommand*\?[1]{\leavevmode\hbox{\normalfont\ttfamily#1}}
\makeatother

\begin{document}

\hbox to\hsize{\hss Ruixi Zhang, November~1, 2021}

\section{About expressibility of \<$X$>\?{sp} in \TeX's units}

In this section, we discuss general input dimension (perhaps in other units).
Again, $X=\lfloor N\delta\rfloor$ for some integer $N\ge0$ and $\delta=P/Q$ is
some unit conversion factor.
Here, we write the fraction $P/Q$ in its lowest term; that is, $\gcd(P,Q)=1$.
Also, since all engine-defined conversion factor $\delta>1$, we assume $P>Q>0$.

Although $N$ can hit every (nonnegative) integer, $X$~cannot. To see why,
write
\[
X=\left\lfloor N\times\frac{P}{Q}\right\rfloor,
\]
and let $N$ run through $1$,~$2$, $3$,~\ldots,~$Q$. Then $X$ must take value
in the range between~$1$ and~$P$ (both ends inclusive).
There are a total of $P$ integers to choose from, but $X$ can only attain $Q$
different values. We conclude that there must exist \<$X$>\?{sp} which is
impossible to be input as \<$n$>\?{.}\<$d_1d_2\ldots d_{16}$>\<unit>.
In fact, for every $P$ consecutive integers,
there are exactly $P-Q$ impossible values.
Let us call such \<$X$>\?{sp} dimension \emph{inexpressible}
when using a particular \<unit>.

\newcommand*\loweruppersp[2]{%
  \begin{array}{@{}l@{}}
  #1\\#2\\
  \end{array}%
}

\medskip

\noindent
{\bfseries Example~1.}
Let \<unit> be~\?{bp}, with its conversion factor $\delta=7227/7200=803/800$.
Then, for every $803$ consecutive integers,
there are exactly $3$ impossible values---%
they are precisely $X\equiv267$, $535$, and $802\pmod{803}$.
In more details, we have
\begin{align*}
\hbox{\?{267sp} is inexpressible:}\quad
&\left\{\loweruppersp{\?{0.00406bp}\implies X=266}{\?{0.00407bp}/\?{0.00408bp}\implies X=268}\right.\\
\hbox{\?{535sp} is inexpressible:}\quad
&\left\{\loweruppersp{\?{0.00813bp}/\?{0.00814bp}\implies X=534}{\?{0.00815bp}\implies X=536}\right.\\
\hbox{\?{802sp} is inexpressible:}\quad
&\left\{\loweruppersp{\?{0.01219bp}\implies X=801}{\?{0.0122bp}/\?{0.01221bp}\implies X=803}\right.
\end{align*}
What about \verb|\maxdimen|, $X=\number\maxdimen=2^{30}-1$\,?
Well, $\number\maxdimen\equiv737\pmod{803}$, and thus \verb|\maxdimen| is
expressible using \?{bp} unit. In fact, either \?{16322.78953bp}
or \?{16322.78954bp} equals \verb|\maxdimen|.

\medskip

\noindent
{\bfseries Example~2.}
Let \<unit> be~\?{pc}, with its conversion factor $\delta=12/1$.
Then, for every $12$ consecutive integers,
there are exactly $11$ impossible values---%
only one value is expressible: $X\equiv0\pmod{12}$.
We have
\[
\hbox{\?{1sp}, \?{2sp}, \?{3sp}, \ldots, \?{11sp} are all inexpressible using \?{pc} unit.}
\]
What about \verb|\maxdimen|, $X=\number\maxdimen=2^{30}-1$\,?
Well, $\number\maxdimen\equiv3\pmod{12}$, and thus \verb|\maxdimen| is
\emph{inexpressible} using \?{pc} unit. In fact,
\[
\left\{\loweruppersp{\?{1365.33333pc}\implies X=1073741820,\quad\hbox{(\?{3sp} short);}}{\?{1365.33334pc}\implies X=1073741832,\quad\hbox{(\?{9sp} over).}}\right.
\]

\end{document}
```

**Comment by RuixiZhang42** [comment]
@blefloch My humble contribution:
```latex
% \iffalse
% The content of this entire code is released under LPPL Version 1.3c.
% To be considered for the l3skip module.
% \fi
%
%    \begin{macrocode}
%<@@=dim>
%    \end{macrocode}
%
%   We first sets up a helper macro \cs[no-index]{@@_tmp:w} which takes two
%   arguments. The first argument is one of the following engine-defined
%   units: |in|,~|pc|, |cm|, |mm|, |bp|, |dd|, |cc|, |nd|, and~|nc|.
%   The second argument is $\frac{1}{2}\delta^{-1}$ in reduced fraction,
%   where $\delta>1$~is the engine-defined conversion factor for each unit.
%   Note that $\delta$~must be strictly larger than~$1$ for the following
%   algorithm to work.
%    \begin{macrocode}
\cs_set_protected:Npn \@@_tmp:w #1#2
  {
    \cs_new:cpn { dim_to_decimal_in_ #1 :n } ##1
      {
        \dim_to_decimal:n
          {
%    \end{macrocode}
%   We need different formulae depending on whether the user input dimension
%   is negative or not. For speed, we use \cs{if_dim:w}.
%    \begin{macrocode}
            \if_dim:w \@@_eval:w ##1 \@@_eval_end: < \c_zero_dim
%    \end{macrocode}
%   For negative dimension (internally represented as $X$\,sp), the formula
%   is $(2X-1)\times(\frac{1}{2}\delta^{-1})$. The intermediate step doubles
%   the dimension~$X$. To avoid underflow at this step, we must invoke
%   \cs[no-index]{__int_eval:w}.
%    \begin{macrocode}
              \__int_eval:w
                ( 2 * \@@_eval:w ##1 \@@_eval_end: - 1 ) * #2
              \__int_eval_end:
            \else:
%    \end{macrocode}
%   For nonnegative dimension, the formula
%   is $(2X+1)\times(\frac{1}{2}\delta^{-1})$. The intermediate step doubles
%   the dimension~$X$. To avoid overflow at this step, we must invoke
%   \cs[no-index]{__int_eval:w} as well.
%    \begin{macrocode}
              \__int_eval:w
                ( 2 * \@@_eval:w ##1 \@@_eval_end: + 1 ) * #2
              \__int_eval_end:
            \fi:
%    \end{macrocode}
%   Now we append~|sp| to the completed integer to finish the dimension.
%    \begin{macrocode}
            sp
          }
      }
  }
%    \end{macrocode}
%
% \begin{macro}[EXP]{\dim_to_decimal_in_in:n}
% \begin{macro}[EXP]{\dim_to_decimal_in_pc:n}
% \begin{macro}[EXP]{\dim_to_decimal_in_cm:n}
% \begin{macro}[EXP]{\dim_to_decimal_in_mm:n}
% \begin{macro}[EXP]{\dim_to_decimal_in_bp:n}
% \begin{macro}[EXP]{\dim_to_decimal_in_dd:n}
% \begin{macro}[EXP]{\dim_to_decimal_in_cc:n}
% \begin{macro}[EXP]{\dim_to_decimal_in_nd:n}
% \begin{macro}[EXP]{\dim_to_decimal_in_nc:n}
%   Conversions to other units are now coded.
%   Consult the pdf\/\TeX{} source for each conversion factor~$\delta$.
%   The factor $\frac{1}{2}\delta^{-1}$ are all hand-coded
%   for accuracy (and speed).
%    \begin{macrocode}
\@@_tmp:w { in } {   50 /  7227 } % delta=7227/100
\@@_tmp:w { pc } {    1 /    24 } % delta=12/1
\@@_tmp:w { cm } {  127 /  7227 } % delta=7227/254
\@@_tmp:w { mm } { 1270 /  7227 } % delta=7227/2540
\@@_tmp:w { bp } {  400 /   803 } % delta=7227/7200
\@@_tmp:w { dd } { 1157 /  2476 } % delta=1238/1157
\@@_tmp:w { cc } { 1157 / 29712 } % delta=14856/1157
\@@_tmp:w { nd } {  321 /   685 } % delta=685/642
\@@_tmp:w { nc } {  107 /  2740 } % delta=1370/107
%    \end{macrocode}
% \end{macro}
% \end{macro}
% \end{macro}
% \end{macro}
% \end{macro}
% \end{macro}
% \end{macro}
% \end{macro}
% \end{macro}
%
```
I&nbsp;will work on `\dim_to_decimal_in_unit:nn` next.

**Comment by jfbu** [comment]
About the issues raised by the `ex` and `em` or more general dimensions not necessarily `>1pt` I have now given it some thought [there](https://github.com/jfbu/texdimens/issues/2#issuecomment-957921157) as a comment to the corresponding potential enhancement of my texdimens package. The code indeed has to branch whether `>1pt̀` and `<1pt`, each range having its peculiarities: for the former this is the impossibility to represent exactly all dimensions in the given unit (on whose various aspects my package texdimens focused so far), and for the latter the issue is about technical difficulties to handle large dimensions in a guaranteed overflow-free manner throughout the whole chain. As we want this expandably we have to carefully think about the usage of `\numexpr` and `\dimexpr` and in my referred-to comment I think I have obtained the way to organize this succesfully.

**Comment by RuixiZhang42** [comment]
@blefloch Here is a slightly faster version to code all 9 `\dim_to_decimal_in_in:n` and friends, by evaluating `\__dim_eval:w ##1 \__dim_eval_end:` only once (instead of twice). It uses an auxiliary function `\__dim_to_decimal_aux:w` to handle the sign of user input.<sup>[1]</sup> This auxiliary can be useful when we extend these 9 special algorithms to handle the general `\dim_to_decimal_in_unit:nn` case.
```latex
%    \begin{macrocode}
%<@@=dim>
\cs_set_protected:Npn \@@_tmp:w #1#2
  {
    \cs_new:cpn { dim_to_decimal_in_ #1 :n } ##1
      {
        \exp_after:wN \@@_to_decimal_aux:w
          \int_value:w \@@_eval:w ##1 \@@_eval_end: ; #2 ;
      }
  }
%    \end{macrocode}
%
% \begin{macro}[EXP]{\@@_to_decimal_aux:w}
%   The tokens after \cs{@@_to_decimal_aux:w} shall have the following form:
%   |<number>;<half of delta inverse>;|, where |<number>| represents the
%   input dimension in |sp| unit.
%   If |<number>| is positive, then |#1| is its leading digit and |#2|
%   (possibly empty) is all the remaining digits;
%   If |<number>| is zero, then |#1| is~|0|$_{12}$ and |#2| is empty;
%   If |<number>| is negative, then |#1| is its sign~|-|$_{12}$ and |#2|
%   is all its digits.
%   In all three cases, |#1#2| is the original |<number>|. We can use |#1|
%   to decide whether to use the |-1| formula or the |+1| formula.
%    \begin{macrocode}
\cs_new:Npn \@@_to_decimal_aux:w #1#2;#3;
  {
    \dim_to_decimal:n
      {
%    \end{macrocode}
%   We need different formulae depending on whether the user input dimension
%   is negative or not.
%   For negative dimension (internally represented as $X$\,sp), the formula
%   is $(2X-1)\times(\frac{1}{2}\delta^{-1})$.
%   For nonnegative dimension, the formula
%   is $(2X+1)\times(\frac{1}{2}\delta^{-1})$.
%   The intermediate step doubles the dimension~$X$.
%   To avoid overflow, we must invoke \cs[no-index]{__int_eval:w}.
%    \begin{macrocode}
        \__int_eval:w
          ( 2 * #1#2 \if:w #1 - - \else: + \fi: 1 ) * #3
        \__int_eval_end:
%    \end{macrocode}
%   Now we append~|sp| to finish the dimension specification.
%    \begin{macrocode}
        sp
      }
  }
%    \end{macrocode}
% \end{macro}
%
% ... ... ...
%    \begin{macrocode}
\@@_tmp:w { in } {   50 /  7227 } % delta=7227/100
% ... ... ...
\@@_tmp:w { nc } {  107 /  2740 } % delta=1370/107
%    \end{macrocode}
% ... ... ...
```
<sup>[1]</sup>&nbsp;The code is largely inspired by [`texdimens`](https://github.com/jfbu/texdimens), admittedly less efficient (but easier to understand I&nbsp;hope).

**Comment by RuixiZhang42** [comment]
@blefloch Here is a brand new `\dim_to_decimal_in_unit:nn` as promised. It covers *any arbitrary units*, as long as the unit is a legal TeX dimension. More specifically:
- If the unit is 0pt (`unit=0`), then a “division by zero” error is raised.
- All units between 1.00001pt and 16383.99999pt (`unit=65537..1073741823`) are handled with a single auxiliary `\__dim_to_decimal_aux:w`, the same auxiliary for `\dim_to_decimal_in_in:n`, `\dim_to_decimal_in_cm:n`, etc.
- If the unit is 1pt (`unit=65536`), then we call `\dim_to_decimal:n`.
- All units between 0.00001pt and 0.99999pt (`unit=1..65535`) are carefully handled to avoid arithmetic overflow.
- All *negative* units are also handled, respecting the fact that TeX truncates toward zero while rounds away from zero.

Three showcases, followed by complete code.

### Showcase&nbsp;1: Improved accuracy

```latex
\newdimen\myunit
\font\1=cmr10 at 3.25mm
\1 % 1em should now equal 606021sp

%%% unit>1pt
\myunit=\dimexpr1em/7\relax

% current kernel implementation:
\dim_to_decimal_in_unit:nn{12.6em}{1em/7}       % gives 88.20047
\ifdim12.6em=88.20047\myunit TRUE\else FALSE\fi % gives FALSE

% proposed new implementation:
\dim_to_decimal_in_unit:nn{12.6em}{1em/7}       % gives 88.20049
\ifdim12.6em=88.20049\myunit TRUE\else FALSE\fi % gives TRUE

%%% unit<1pt
\myunit=\dimexpr1em/50\relax

% current kernel implementation:
\dim_to_decimal_in_unit:nn{12.6em}{1em/50}       % gives 630.02211
\ifdim12.6em=630.02211\myunit TRUE\else FALSE\fi % gives FALSE

% proposed new implementation:
\dim_to_decimal_in_unit:nn{12.6em}{1em/50}       % gives 630.02213
\ifdim12.6em=630.02213\myunit TRUE\else FALSE\fi % gives TRUE
```

### Showcase&nbsp;2: No more `! Arithmetic overflow` error

```latex
\dim_to_decimal_in_unit:nn{\maxdimen}{1pt-1sp}
% current kernel implementation fails with ! Arithmetic overflow.
% proposed new implementation gives 16384.25

\dim_to_decimal_in_unit:nn{5651.4033pt}{3sp}
% current kernel implementation fails with ! Arithmetic overflow.
% proposed new implementation gives 123456789

\dim_to_decimal_in_unit:nn{\maxdimen}{1sp}
% current kernel implementation fails with ! Arithmetic overflow.
% proposed new implementation gives 1073741823
```

### Showcase&nbsp;3: Negative dimensions

```latex
% proposed new implementation:
\dim_to_decimal_in_unit:nn{ \maxdimen}{1pt/3-7sp} % gives  49168.50551
\dim_to_decimal_in_unit:nn{-\maxdimen}{1pt/3-7sp} % gives -49168.50551
\dim_to_decimal_in_unit:nn{ \maxdimen}{7sp-1pt/3} % gives -49168.50551
\dim_to_decimal_in_unit:nn{-\maxdimen}{7sp-1pt/3} % gives  49168.50551
```

### Code for consideration

```latex
%
%   This one is for the l3msg module.
%
%    \begin{macrocode}
\msg_new:nnn { kernel } { zero-unit }
  { Zero~unit~for~dimexpr2. }
%    \end{macrocode}
%
%
%   Now back to the l3skip module.
%
% \begin{macro}[EXP]{\dim_to_decimal_in_unit:nn}
%    \begin{macrocode}
%<@@=dim>
\cs_new:Npn \dim_to_decimal_in_unit:nn #1#2
  {
    \exp_after:wN \@@_chk_unit:w
      \int_value:w \@@_eval:w #2 \@@_eval_end: ; {#1}
  }
%    \end{macrocode}
% \end{macro}
%
% \begin{macro}[EXP]{\@@_chk_unit:w}
%   The tokens after \cs{@@_chk_unit:w} shall have the following form:
%   |<number2>;{<dimexpr1>}|, where |<number2>| represents |<dimexpr2>| in
%   |sp| unit.
%   If |#1| is~|0|$_{12}$, the ``unit'' |<dimexpr2>| must also be zero.
%   So we throw out a ``division by zero'' error message at this point.
%   Otherwise, if |#1| is~|-|$_{12}$, we shall negate both |<dimexpr1>| and
%   |<dimexpr2>| for later procedures.
%    \begin{macrocode}
\cs_new:Npn \@@_chk_unit:w #1#2;#3
  {
    \token_if_eq_charcode:NNTF #1 0
      { \msg_expandable_error:nn { kernel } { zero-unit } }
      {
        \exp_after:wN \@@_branch_unit:w
          \int_value:w \if:w #1 - - \fi: \@@_eval:w #3 \exp_after:wN ;
          \int_value:w \if:w #1 - - \fi: #1#2 ;
      }
  }
%    \end{macrocode}
% \end{macro}
%
% \begin{macro}[EXP]{\@@_branch_unit:w}
%   The tokens after \cs{@@_branch_unit:w} shall have the following form:
%   |<number1>;<number2>;|, where |<number1>| represents |<dimexpr1>| in
%   |sp| unit (whose sign is taken care of) and |<number2>| represents the
%   absolute value of |<dimexpr2>| in |sp| unit (which is strictly positive).
%
%   As explained, the formulae $(2X\pm1)\times(\frac{1}{2}\delta^{-1})$ work
%   if and only if $\delta=|<number2>|/65536>1$. This corresponds to
%   |<dimexpr2>| strictly larger than 1\,pt in absolute value.
%   In this case, we simply call \cs{@@_to_decimal_aux:w} and supply
%   $\frac{1}{2}\delta^{-1}=32768/|<number2>|$ as |<half of delta inverse>|.
%
%   Otherwise if $|<number2>|=65536$, then |<dimexpr2>| is 1\,pt in absolute
%   value and we call \cs{dim_to_decimal:n} directly.
%
%   Otherwise $0<|<number2>|<65536$ and we shall proceed differently.
%
%   For unit less than 1\,pt, write $n=|<number2>|$, then $\delta=n/65536<1$.
%   The midpoint formulae are not optimal. Let's go back to the inequalities
%   $X\delta^{-1}\le N<(X+1)\delta^{-1}$. Since now $\delta<1$, the bounding
%   interval is wider than~$1$ in length. Consider the ceiling integer
%   $M=\lceil X\delta^{-1}\rceil$, then $X\delta^{-1}\le M<(X+1)\delta^{-1}$,
%   or equivalently $X\le M\delta<X+1$, and thus $\lfloor M\delta\rfloor=X$.
%   The key point here is that we \emph{don't} need to solve for~$N$;
%   in fact, any integer that can reproduce~$X$ (such as~$M$) is good enough.
%   So the algorithm goes like this: (1)~Compute rounding of $X\delta^{-1}$,
%   i.e., $M'=\lfloor X\delta^{-1}+\frac{1}{2}\rfloor$; this $M'$ could be
%   either $M$ or $M-1$. (2)~Check if $\lfloor M'\delta\rfloor=X$, i.e.,
%   whether our candidate $M'$ can reproduce~$X$. If so, then this $M'$ is
%   good enough; if not, then we add one to~$M'$.
%
%   But when $0<n<65536$, we cannot delay the problem of overflow any more.
%   For $X\delta^{-1}=X\times65536/n$, where $X$ can go up to $2^{30}-1$ and
%   $n$ can be as small as~$1$, the result is well over $2^{31}-1$ (largest
%   integer allowed within |\numexpr|).
%   For example, |\dim_to_decimal_in_unit:nn { \maxdimen } { 1sp }|.
%   Here, all inputs are legal, so we should be able to output |1073741823|
%   \emph{without} causing arithmetic overflow.
%
%   As a workaround, let's write $X=qn+r$ with some $q\ge0$ and $0\le r<n$.
%   Then $X\delta^{-1}=65536q+65536r/n$, and so
%   $M'=65536q+\lfloor65536r/n+\frac{1}{2}\rfloor=65536q+R'$.
%   Computing $R'$ will never overflow. If this $R'$ can reproduce~$r$, then
%   it is good enough; otherwise we add one to~$R'$. In the end, we shall
%   output $q+R'/65536$ in decimal.
%
%   Note: $q=\lfloor X/n\rfloor=\lfloor\frac{2X-n}{2n}+\frac{1}{2}\rfloor$
%   represents the ``integer'' part, while $0\le R'\le65536$ represents the
%   ``fractional'' part. (Can $R'=65536$ really happen? Didn't investigate.)
%    \begin{macrocode}
\cs_new:Npn \@@_branch_unit:w #1;#2;
  {
    \int_compare:nNnTF {#2} > { 65536 }
      { \@@_to_decimal_aux:w #1 ; 32768 / #2 ; }
      {
        \int_compare:nNnTF {#2} = { 65536 }
          { \dim_to_decimal:n { #1sp } }
          { \@@_get_quotient:w #1 ; #2 ; }
      }
  }
%    \end{macrocode}
% \end{macro}
%
% \begin{macro}[EXP]{\@@_get_quotient:w}
%   We wish to get the quotient $q$ via rounding of $\frac{2X-n}{2n}$.
%   When $0\le X<n/2$, we have $\frac{2X-n}{2n}<0$. So, strictly speaking,
%   |\numexpr| performs its rounding as
%   $\lceil\frac{2X-n}{2n}-\frac{1}{2}\rceil$, not exactly what we want.
%   However, lucky for us, only $X=0$ makes
%   $\lceil\frac{2X-n}{2n}-\frac{1}{2}\rceil=-1\neq0$ (we want~$0$);
%   all other $0<X<n/2$ make $\lceil\frac{2X-n}{2n}-\frac{1}{2}\rceil=0=q$.
%   Thus, let's filter out $X=0$ early.
%   If $X\neq0$, we extract its sign and leave the sign to the back.
%   The sign does not participate in any calculations (also the code works
%   with positive integers only). The sign is used at the last stages when
%   we parse the decimal output.
%
%   After \cs{@@_get_quotient:w} has done its job, either we have the
%   decimal~|0|, or we have \cs{@@_get_remainder:w} followed by
%   $q$|;|$\lvert X\rvert$|;|$n$|;<sign of X>;|.
%    \begin{macrocode}
\cs_new:Npn \@@_get_quotient:w #1#2;#3;
  {
    \token_if_eq_charcode:NNTF #1 0
      { 0 }
      {
        \token_if_eq_charcode:NNTF #1 -
          {
            \exp_after:wN \@@_get_remainder:w
              \int_use:N \__int_eval:w ( 2 * #2 - #3 ) / ( 2 * #3 ) ;
              #2 ; #3 ; - ;
          }
          {
            \exp_after:wN \@@_get_remainder:w
              \int_use:N \__int_eval:w ( 2 * #1#2 - #3 ) / ( 2 * #3 ) ;
              #1#2 ; #3 ; ;
          }
      }
  }
%    \end{macrocode}
% \end{macro}
%
% \begin{macro}[EXP]{\@@_get_remainder:w}
%   \cs{@@_get_remainder:w} does not need to read the sign.
%   After finding the remainder~$r$, the number~$\lvert X\rvert$ is no longer
%   needed. We should then have \cs{@@_convert_remainder:w} followed by
%   $r$|;|$n$|;|$q$|;<sign of X>;|.
%    \begin{macrocode}
\cs_new:Npn \@@_get_remainder:w #1;#2;#3;
  {
    \exp_after:wN \@@_convert_remainder:w
      \int_use:N \__int_eval:w #2 - #1 * #3 ;
      #3 ; #1 ;
  }
%    \end{macrocode}
% \end{macro}
%
% \begin{macro}[EXP]{\@@_convert_remainder:w}
%   This is trivial. We compute $R'=\lfloor65536r/n+\frac{1}{2}\rfloor$,
%   then leave \cs{@@_test_candidate:w} followed by
%   $R'$|;|$r$|;|$n$|;|$q$|;<sign of X>;|.
%    \begin{macrocode}
\cs_new:Npn \@@_convert_remainder:w #1;#2;
  {
    \exp_after:wN \@@_test_candidate:w
      \int_use:N \__int_eval:w #1 * 65536 / #2 ;
      #1 ; #2 ;
  }
%    \end{macrocode}
% \end{macro}
%
% \begin{macro}[EXP]{\@@_test_candidate:w}
%   Now the fun part: We take $R'$, $r$ and~$n$ to test whether
%   $r=\lfloor R'\delta\rfloor$. This is done as a dimension comparison.
%   The left-hand side, $r$, is simply |r sp|. The right-hand side,
%   $\lfloor R'\delta\rfloor$, is exactly |<R' as decimal><dimen = n sp>|.
%   If the result is true, then we've found~$R'$;
%   otherwise we add one to~$R'$.
%   After this step, $r$ and~$n$ are no longer needed. We should then have
%   \cs{@@_parse_decimal:w} followed by $R'$|;|$q$|;<sign of X>;|.
%    \begin{macrocode}
\cs_new:Npn \@@_test_candidate:w #1;#2;#3;
  {
    \dim_compare:nNnTF { #2sp } =
      { \dim_to_decimal:n { #1sp } \@@_eval:w #3sp \@@_eval_end: }
      { \@@_parse_decimal:w #1 ; }
      {
        \exp_after:wN \@@_parse_decimal:w
          \int_use:N \__int_eval:w #1 + 1 ;
      }
  }
%    \end{macrocode}
% \end{macro}
%
% \begin{macro}[EXP]{\@@_parse_decimal:w, \@@_parse_decimal_aux:w}
%   The Grand Finale: We sum $q$ and $R'/65536$ together, and negate the
%   result if necessary. These are all done expandably.
%   If $0<R'/65536<1$, the integer summation is naturally terminated at the
%   decimal point. If $R'/65536=0$ (or~$1$?), the summation is terminated
%   at the semicolon. The auxiliary function \cs{@@_parse_decimal_aux:w}
%   takes care of both cases.
%    \begin{macrocode}
\cs_new:Npn \@@_parse_decimal:w #1;#2;#3;
  {
    \exp_after:wN \@@_parse_decimal_aux:w
      \int_value:w #3 \__int_eval:w #2 + \dim_to_decimal:n { #1sp } ;
  }
\cs_new:Npn \@@_parse_decimal_aux:w #1; {#1}
%    \end{macrocode}
% \end{macro}
%
```

**Comment by josephwright** [comment]
> The discussions here have impact on real-world applications. I noticed one such anomaly when maintaining the `ctex` bundle. The `ctex` document classes offer a `10.5bp` font size for body text, known as 五号 in simplified Chinese (five hao). The natural leading is 1.2 times the font size, i.e., `12.6bp`.

I'm reading back and hoping I can resolve the issue here, but I'm wondering about this statement. If the class is accepting `10.5bp`, then the baseline skip should be set to `1.2\<some-dim>`, so I don't see how the conversion back to `bp` impacts. (I suspect I might need to add something to the docs pointing out that `\dim_to_decimal_in_bp:n` is for places we *have* to convert, such as the backend where we need to pass data to PostScript/PDF). So I wonder if you could explain a bit more how things are being used: I'm worrying I've missed a use case!

**Comment by josephwright** [comment]
With hindsight, I guess the docs for `\dim_to_decimal_in_bp:n` should have said what it was (intended) for! At the backend level, one needs to work with PostScript or PDF or ..., i.e. things that regard 'points' as 1/72 in. As with many backend conversions, we need to be 'good enough': experience from the classical `.def` files shows that sound rounding is 'acceptable' as the backends themselves have their own 'features'. 

However, that does not mean that improvements are not welcome, only that we have to balance off effort and also performance (`\dim_to_decimal_in_bp:n` is needed a lot in `l3draw`, for example). I think looking over @RuixiZhang42's code I can adjust what we have without too much cost at point of use so I will look to integrate.

**Comment by RuixiZhang42** [comment]
@josephwright I&nbsp;find all 5 **but one** replacement of `\int_use:N \__int_eval:w` to `\int_eval:w` strange. In the first 4 instances, of course three `\exp_after:wN`&rsquo;s were necessary. The last instance was in `\@@_test_candidate:w`, false branch, where only one `\exp_after:wN` was present (which surprisingly didn&rsquo;t raise any error!). On closer inspection, I&nbsp;think it&rsquo;s perfectly safe to drop that single `\exp_after:wN` too:
```latex
\cs_new:Npn \@@_test_candidate:w #1;#2;#3;
  {
    \dim_compare:nNnTF { #2sp } =
      { \dim_to_decimal:n { #1sp } \@@_eval:w #3sp \@@_eval_end: }
      { \@@_parse_decimal:w #1 ; }
      { \@@_parse_decimal:w \int_eval:w #1 + 1 ; } % <- no need to expand \int_eval:w at all!
  }
```
No error, because the stuff between `\@@_parse_decimal:w` and the first semicolon&nbsp;`;` is appended with `sp` and then passed to `\dim_to_decimal:n`. Essentially, the final step is doing an
```latex
% Expansion of \@@_parse_decimal:w
... \expandafter \__dim_to_decimal:w \the \dimexpr \int_eval:w <R'>+1 sp \relax ...
```

**Comment by josephwright** [comment]
@RuixiZhang42 The standard `expl3` rules for internals apply, so we can't use `\__int_eval:w` in the `dim` module. (We also can't use `\int_use:N \__int_eval:w`, as `\int_use:N` takes an `N` type argument which should be a `dim`.) So the options are

- Use an alternative documented interface
- Make a local copy of the primitives and use that

Of the two, we usually favour the first approach unless there is a strong reason to do otherwise.

Looking back, I realise that we don't document that `\int_eval:w` needs exactly two expansions, but we do for `\int_eval:n`. So I will look again at what is required/clearest.
